### PR TITLE
Tuple unpacking now works for `for` vars

### PR DESCRIPTION
**Pull request by nc-x**
Implements and closes https://github.com/nim-lang/Nim/issues/7486

~~Todo:~~
~~1) refactor~~
~~2) add nested tuple unpacking support for var/let/const/for~~ - Later in another PR

### CODE REVIEW

File: compiler/parser.nim [hunk 1]
```diff
@@ -1154,23 +1154,48 @@ proc parseTypeDescKAux(p: var TParser, kind: TNodeKind,
...
+      addSon(tup, a)
+  addSon(result, tup)
+  eat(p, tkParRi)
   eat(p, tkIn)
   addSon(result, parseExpr(p))
   colcom(p, result)
   addSon(result, parseStmt(p))
 
+proc parseFor(p: var TParser): PNode =
+  #| forStmt = 'for' (identWithPragma ^+ comma) 'in' expr colcom stmt
```
Comment (nc-x) [inline compiler/parser.nim]: help needed fixing this
cc @Araq

**Comment by nc-x** [comment]
Unrelated failure

**Comment by krux02** [comment]
Generally I don't like new features, because they might cause problems that are impossible to predict. And I would like to focus the nim development towards stability. But for this feature I might make an exception, because I really like it. Tuple unpacking is something that I already did, but it was always very limited and a bit awkward. 

But you need to add documentation for this feature. Explain in a tutorial the difference between `for a,b  in x` and `for (a,b) in x`. And explain what iterator is chosen it what situations. For example I would mention that tuple unpacking is supported, with `()` and without `()` at top level. But when there is a (), then the compiler tries to find an iterator with the name `items`, and when there is no `()` at top level the compiler tries to find an iterator with the name `pairs`.

I used this syntax for tuple unpacking:

```nim
let t1 = (1, 2, 3)
let t2 = (11, 32, 83)
let s = @[t1, t2]

for a, b, c in s.items:
  echo a, b, c
```

But I don't like it. I think it should be deprecated in favor of `for (a, b, c) in s.items:` or just `for (a, b, c) in s:`.


Also you should at least mention the `mitems` and `mpairs` variants. Can they be supported with unpacking?

**Comment by nc-x** [comment]
@krux02 
The rules are same as they used to be
```nim
for i, (a, b, c) in x              #=>  calls pairs() on x; pairs returns (index, (val1, val2, val3)) which gets unpacked
for (a, b, c) in x                 #=>  calls items() on x; items returns (val1, val2, val3) which gets unpacked
```

> Also you should at least mention the mitems and mpairs variants. Can they be supported with unpacking?
  
They are not supported yet. (Mainly because I don't know how to create a PType - tyVar(tyInt `or whatever`)).
    
----
Apart from that, this PR currently has issues wrt error messages -
```
var t1 = (1, 2, 3)
var t2 = (11, 32, 83)
var s = @[t1, t2]

for (a,b,c) in s.pairs():
  echo a
```
gives 
```
t.nim(5, 5) Error: required number of tuple variables: 2; given: 3
```
---
  
Unfortunately I won't have time for the next few weeks, so either this PR needs to wait, or someone else needs to take over from here.

**Comment by nc-x** [comment]
Ready for review.

**Comment by nc-x** [comment]
@krux02 
> I used this syntax for tuple unpacking:
```
let t1 = (1, 2, 3)
let t2 = (11, 32, 83)
let s = @[t1, t2]

for a, b, c in s.items:
  echo a, b, c
```
> But I don't like it. I think it should be deprecated in favor of for (a, b, c) in s.items: or just for (a, b, c) in s:.

That is how `for i, j in x.pairs()` work. The compiler implicitly unpacks 1st level of tuple. Deprecating that syntax would mean that parenthesis would have to be used with `.pairs()` like `for (i, j) in x.pairs()`
or `for (i, (a,b,c)) in x.pairs()` (in case it returns tuples).
  
If that is desired, I can send a PR for the deprecation. Otherwise, I think this PR is ready.

**Comment by nc-x** [comment]
ping @Araq @krux02

**Comment by Araq** [comment]
@narimiran Please rebase so that we know if it breaks our selected Nimble packages.

**Comment by nc-x** [comment]
Merged devel into my branch.

**Comment by nc-x** [comment]
ping @Araq 
The CI's are green.

**Comment by Araq** [comment]
Thank you for your patience.

**Pull request merged**